consensus/hotstuff/safetyrules/safety_rules_test.go

```diff
@@ -1,165 +1,526 @@
+	require.True(s.T(), model.IsNoVoteError(err))
+	s.persister.AssertNotCalled(s.T(), "PutSafetyData")
+}
 
-	_, err := voter.ProduceVote(block, curView)
-	require.Error(t, err)
-	require.Contains(t, err.Error(), "expecting block for current view")
-	require.False(t, model.IsNoVoteError(err))
+// TestProduceVote_InvalidVoterIdentity tests that no vote is created if there was an exception retrieving voter identity
+func (s *SafetyRulesTestSuite) TestProduceVote_InvalidVoterIdentity() {
+	*s.committee = mocks.DynamicCommittee{}
+	s.committee.On("Self").Return(s.ourIdentity.NodeID)
+	exception := errors.New("invalid-signer-identity")
+	s.committee.On("IdentityByBlock", s.proposal.Block.BlockID, s.proposal.Block.ProposerID).Return(s.proposerIdentity, nil).Maybe()
+	s.committee.On("IdentityByBlock", s.proposal.Block.BlockID, s.ourIdentity.NodeID).Return(nil, exception).Once()
+
+	vote, err := s.safety.ProduceVote(s.proposal, s.proposal.Block.View)
+	require.Nil(s.T(), vote)
+	require.ErrorIs(s.T(), err, exception)
+	s.persister.AssertNotCalled(s.T(), "PutSafetyData")
 }
 
-func testEqualLastVotedView(t *testing.T) {
-	// curView == lastVotedView
-	blockView, curView, lastVotedView, isBlockSafe, isCommitteeMember := uint64(3), uint64(3), uint64(3), true, true
+// TestProduceVote_CreateVoteException tests that no vote is created if vote creation raised an exception
+func (s *SafetyRulesTestSuite) TestProduceVote_CreateVoteException() {
+	exception := errors.New("create-vote-exception")
+	s.signer.On("CreateVote", s.proposal.Block).Return(nil, exception).Once()
+	vote, err := s.safety.ProduceVote(s.proposal, s.proposal.Block.View)
+	require.Nil(s.T(), vote)
+	require.ErrorIs(s.T(), err, exception)
+	s.persister.AssertNotCalled(s.T(), "PutSafetyData")
+}
 
-	// create voter
-	block, _, voter := createVoter(blockView, lastVotedView, isBlockSafe, isCommitteeMember)
+// TestProduceVote_PersistStateException tests that no vote is created if persisting state failed
+func (s *SafetyRulesTestSuite) TestProduceVote_PersistStateException() {
+	exception := errors.New("persister-exception")
+	s.persister.On("PutSafetyData", mock.Anything).Return(exception)
 
-	_, err := voter.ProduceVote(block, curView)
-	require.Error(t, err)
-	require.Contains(t, err.Error(), "must be larger than the last voted view")
-	require.False(t, model.IsNoVoteError(err))
+	vote := makeVote(s.proposal.Block)
+	s.signer.On("CreateVote", s.proposal.Block).Return(vote, nil).Once()
+	vote, err := s.safety.ProduceVote(s.proposal, s.proposal.Block.View)
+	require.Nil(s.T(), vote)
+	require.ErrorIs(s.T(), err, exception)
 }
 
-func testBelowLastVotedView(t *testing.T) {
-	// curView < lastVotedView
-	blockView, curView, lastVotedView, isBlockSafe, isCommitteeMember := uint64(3), uint64(3), uint64(4), true, true
+// TestProduceVote_VotingOnUnsafeProposal tests different scenarios where we try to vote on unsafe blocks
+func (s *SafetyRulesTestSuite) TestProduceVote_VotingOnUnsafeProposal() {
```

Comment:
## ❓ (potentially missing test?)

Do we have a test that verifies the following rule:
* If a QC for the previous view is included in a block, the TC _must_ be nil? Otherwise, the block violates protocol rules.
* violating this rule should lead to an exception, because we expect the blocks to be already verified. 

Essentially rule 2 and 3 [here](https://github.com/onflow/flow-go/blob/3f1674597fdb15a2822f42db3e0101e2e8e136b4/consensus/hotstuff/safetyrules/safety_rules.go#L205-L216): 
> Happy Path [...] to prevent resource-exhaustion attacks, we require that no TC is included as part of the proof. 
> Following the Happy Path has priority over following the Recovery Path

Reply:
This test case is included in proposal [validation](https://github.com/onflow/flow-go/pull/2456/files#diff-c85759ef2971fa1acf44671dfb9dbf30e7f3485d3920d39686c2fdff08527518R403) only. I will add same test case here.

Reply:
We are discussing the convention:
> If a QC for the previous view is included in a block, the TC must be nil

From my perspective, there are two angles to this discussion:
1. Will the proofs still hold, if `SafetyRules` does not enforce this convention?
2. Is our implementation of `SafetyRules` enforcing this convention. 

How I understand your argument, you are focusing on 1. Here, the answer is "yes, SafetyRules does not need to enforce this convention".  Hence, one could conclude 

In comparison, I am more focusing on 2. I think our tests for safety rules should be tailored towards our specific implementation of `SafetyRules`. 
* At the moment, our `SafetyRules` documentation states that it enforces this convention: https://github.com/onflow/flow-go/blob/b6f2aa302e3709d902fa1b61df18f250c6b3b218/consensus/hotstuff/safetyrules/safety_rules.go#L206-L209 Rules 1. and 2. directly imply that a block that violates our convention should not pass `validateEvidenceForEnteringView`
* Also the implementation has the specific code to reject blocks that violate the condition: https://github.com/onflow/flow-go/blob/b6f2aa302e3709d902fa1b61df18f250c6b3b218/consensus/hotstuff/safetyrules/safety_rules.go#L237-L239

In summary, I conclude that `SafetyRules` aims to enforce the convention.  Hence, I think the tests should verify the expected behaviour.

Reply:
Btw, the convention is not only w.r.t. to blocks but also applies to timeouts. I would prefer to have tests, where we verify that violating this convention results in an error (not only for blocks but also for timeouts). 
Having said this, this is a soft preference. I am also ok with your position :-)